File: apps/ruby-sdk/lib/firecrawl/models/crawl_options.rb
# frozen_string_literal: true

module Firecrawl
  module Models
    # Options for crawling a website.
    class CrawlOptions
      FIELDS = %i[
        prompt exclude_paths include_paths max_discovery_depth sitemap
        ignore_query_parameters deduplicate_similar_urls limit
        crawl_entire_domain allow_external_links allow_subdomains
        delay max_concurrency webhook scrape_options regex_on_full_url
        zero_data_retention integration
      ].freeze

      attr_reader(*FIELDS)

      def initialize(**kwargs)
        FIELDS.each { |f| instance_variable_set(:"@#{f}", kwargs[f]) }
      end

      def to_h
        h = {
          "prompt" => prompt,
          "excludePaths" => exclude_paths,
          "includePaths" => include_paths,
          "maxDiscoveryDepth" => max_discovery_depth,
          "sitemap" => sitemap,
          "ignoreQueryParameters" => ignore_query_parameters,
          "deduplicateSimilarURLs" => deduplicate_similar_urls,
          "limit" => limit,
          "crawlEntireDomain" => crawl_entire_domain,
          "allowExternalLinks" => allow_external_links,
          "allowSubdomains" => allow_subdomains,
          "delay" => delay,
          "maxConcurrency" => max_concurrency,
          "webhook" => serialize_webhook,
          "scrapeOptions" => scrape_options&.to_h,
          "regexOnFullURL" => regex_on_full_url,
          "zeroDataRetention" => zero_data_retention,
          "integration" => integration,
        }.compact
        h
      end

      private

      def serialize_webhook
        return webhook if webhook.is_a?(String) || webhook.is_a?(Hash)
        webhook&.to_h
      end
    end
  end
end

File: apps/dot-net-sdk/Firecrawl/Models/CrawlOptions.cs
using System.Text.Json.Serialization;

namespace Firecrawl.Models;

/// <summary>
/// Configuration options for crawling a website.
/// </summary>
public class CrawlOptions
{
    [JsonPropertyName("prompt")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Prompt { get; set; }

    [JsonPropertyName("excludePaths")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public List<string>? ExcludePaths { get; set; }

    [JsonPropertyName("includePaths")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public List<string>? IncludePaths { get; set; }

    [JsonPropertyName("maxDiscoveryDepth")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public int? MaxDiscoveryDepth { get; set; }

    [JsonPropertyName("sitemap")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Sitemap { get; set; }

    [JsonPropertyName("ignoreQueryParameters")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public bool? IgnoreQueryParameters { get; set; }

    [JsonPropertyName("deduplicateSimilarURLs")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public bool? DeduplicateSimilarURLs { get; set; }

    [JsonPropertyName("limit")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public int? Limit { get; set; }

    [JsonPropertyName("crawlEntireDomain")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public bool? CrawlEntireDomain { get; set; }

    [JsonPropertyName("allowExternalLinks")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public bool? AllowExternalLinks { get; set; }

    [JsonPropertyName("allowSubdomains")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public bool? AllowSubdomains { get; set; }

    [JsonPropertyName("delay")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public int? Delay { get; set; }

    [JsonPropertyName("maxConcurrency")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public int? MaxConcurrency { get; set; }

    [JsonPropertyName("webhook")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public object? Webhook { get; set; }

    [JsonPropertyName("scrapeOptions")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public ScrapeOptions? ScrapeOptions { get; set; }

    [JsonPropertyName("regexOnFullURL")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public bool? RegexOnFullURL { get; set; }

    [JsonPropertyName("zeroDataRetention")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public bool? ZeroDataRetention { get; set; }

    [JsonPropertyName("integration")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Integration { get; set; }
}
